Objective-C/CBLReadOnlyDictionary.mm

```diff
@@ -0,0 +1,194 @@
+//
+//  CBLReadOnlyDictionary.m
+//  CouchbaseLite
+//
+//  Created by Pasin Suriyentrakorn on 4/11/17.
+//  Copyright © 2017 Couchbase. All rights reserved.
+//
+
+#import "CBLReadOnlyDictionary.h"
+#import "CBLData.h"
+#import "CBLDocument+Internal.h"
+#import "CBLInternal.h"
+#import "CBLJSON.h"
+#import "CBLStringBytes.h"
+#import "CBLSharedKeys.hh"
+
+
+@implementation CBLReadOnlyDictionary {
+    CBLFLDict* _data;
+    FLDict _dict;
+    cbl::SharedKeys _sharedKeys;
+}
+
+@synthesize data=_data;
+
+- /* internal */ (instancetype) initWithFleeceData: (nullable CBLFLDict*)data {
+    self = [super init];
+    if (self) {
+        self.data = data;
+    }
+    return self;
+}
+
+
+- (NSUInteger) count {
+    return FLDict_Count(_dict);
+}
+
+
+- (nullable id) objectForKey: (NSString*)key {
+    return [self fleeceValueToObject: [self fleeceValueForKey: key]];
+}
+
+
+- (BOOL) booleanForKey: (NSString*)key {
+    return FLValue_AsBool([self fleeceValueForKey: key]);
+}
+
+
+- (NSInteger) integerForKey: (NSString*)key {
+    return (NSInteger)FLValue_AsInt([self fleeceValueForKey: key]);
+}
+
+
+- (float) floatForKey: (NSString*)key {
+    return FLValue_AsFloat([self fleeceValueForKey: key]);
+}
+
+
+- (double) doubleForKey: (NSString*)key {
+    return FLValue_AsDouble([self fleeceValueForKey: key]);
+}
+
+
+- (nullable NSString*) stringForKey: (NSString*)key {
+    return $castIf(NSString, [self objectForKey: key]);
+}
+
+
+- (nullable NSNumber*) numberForKey: (NSString*)key {
+    return $castIf(NSNumber, [self objectForKey: key]);
+}
+
+
+- (nullable NSDate*) dateForKey: (NSString*)key {
+    return [CBLJSON dateWithJSONObject: [self stringForKey: key]];
+}
+
+
+- (nullable CBLBlob*) blobForKey: (NSString*)key {
+    return $castIf(CBLBlob, [self objectForKey: key]);
+}
+
+
+- (nullable CBLReadOnlySubdocument*) subdocumentForKey: (NSString*)key {
+    return $castIf(CBLReadOnlySubdocument, [self objectForKey: key]);
+}
+
+
+- (nullable CBLReadOnlyArray*) arrayForKey: (NSString*)key {
+    return $castIf(CBLReadOnlyArray, [self objectForKey: key]);
+}
+
+
+- (BOOL) containsObjectForKey: (NSString*)key {
+    FLValueType type = FLValue_GetType([self fleeceValueForKey: key]);
+    return type != kFLUndefined;
+}
+
+
+- (NSArray*) allKeys {
+    NSMutableArray* keys = [NSMutableArray array];
+    if (_dict != nullptr) {
+        FLDictIterator iter;
```

Comment:
`iter` should be released in general.

I believe FLDictIterator automatically release itself when it reaches at the end.